### PR TITLE
dbt-materialize: audit mz_generate_name, api relations, make database aware

### PR DESCRIPTION
**Pull request by ahelium**
- Deprecate mz_generate_name function in favor of internal dbt functionality 
- Make dbt-materialize database aware
- Audit the way relations are dropped pre create
- Adjust catalog query to include materialize specific attributes

### Motivation

This PR fixes previously unreported bugs.

- This [PR](https://github.com/MaterializeInc/materialize/pull/13706/files) introduced cloud specific code into the adapter. Specifically in the `list_relations_without_caching` and `catalog` macros. I adjusted those macros to work with both the binary and cloud. This _might_ prolong the need for us to check the version of materialize we are running against.   

  A question: I diverged from using pg_catalog here because mz_catalog has a better representation of the resources (sources, sinks, indexes, (in the case of cloud) materialized views). Is this ok? Is there a better/different query we should be using? 

- Old Relations had different types than New Relations, and so were not getting correctly dropped in the custom materialization macros: for example, adding the below to the source custom [materialization](https://github.com/MaterializeInc/materialize/pull/14036/files#diff-bde408dfb29db5e4f97c46f1d4342c864168143043d78cc826d61dfc55611aa6) 
  ```
  {{ log("OLD Relation: " ~ old_relation.type, info=true) }}
  {{ log("NEW Relation: " ~ target_relation.type, info=true) }}
  ```
  And running a source model: 
  ```
  18:40:04  1 of 3 START source model public.market_orders_raw ............................. [RUN]
  CLASS TYPE <class 'dbt.adapters.materialize.relation.MaterializeRelation'>
  18:40:04  OLD Relation: table
  18:40:04  NEW Relation: source
  ```
  The [get_relation](https://github.com/dbt-labs/dbt-core/blob/main/core/dbt/adapters/base/impl.py#L745) function runs the `list_relations_without_caching` on first run, and caches the results for dbt to more quickly grab next time. The cached relations were getting stored with BaseRelation type value, not MaterializeRelationType specific values we need to correctly drop sources, sinks, etc.  Seeds were also not getting dropped correctly, but now should be. 

### Tips for reviewer

I've included @morsapaes work to use Redpanda instead of Kafka for testing. 
PR for removing mz_generate_name macro from demos/docs is next, should this get approved. 

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.

### CODE REVIEW

**Comment by morsapaes** [comment]
> I diverged from using pg_catalog here because mz_catalog has a better representation of the resources (sources, sinks, indexes, (in the case of cloud) materialized views). Is this ok? Is there a better/different query we should be using?

If you check [the catalog definition](https://github.com/MaterializeInc/materialize/blob/8637d1df65d7ddb6b9f952575c90df24f543e931/src/adapter/src/catalog/builtin.rs#L1945) (maybe you did, sorry!), all `pg_catalog` views are built on top of `mz_catalog` system tables. I like that you switched, it makes things cleaner and easier to parse. Having the dbt documentation show the wrong object types has bothered me since the beginning of times, too!

File: misc/dbt-materialize/dbt/adapters/materialize/impl.py [hunk 1]
```diff
@@ -36,3 +39,34 @@ def _link_cached_relations(self, manifest):
...
         # [0]: https://github.com/dbt-labs/dbt-core/blob/13b18654f03d92eab3f5a9113e526a2a844f145d/plugins/postgres/dbt/adapters/postgres/impl.py#L126-L133
         pass
+
+    def verify_database(self, database):
+        pass
+
+    def list_relations_without_caching(
+        self, schema_relation: MaterializeRelation
+    ) -> List[MaterializeRelation]:
+        # Adapted from: https://github.com/dbt-labs/dbt-snowflake/blob/main/dbt/adapters/snowflake/impl.py#L110
```
Comment (morsapaes) [inline misc/dbt-materialize/dbt/adapters/materialize/impl.py:49]:
```suggestion
```

Comment (morsapaes) [reply misc/dbt-materialize/dbt/adapters/materialize/impl.py:49]: This is pretty much the core implementation: https://github.com/dbt-labs/dbt-core/blob/c270a77552ae9fc66fdfab359d65a8db1307c3f3/core/dbt/adapters/sql/impl.py#L178

**Review by morsapaes — APPROVED** [review]
So many improvements! Since you have tested and noodled on these changes for a while, I just had a look through the code and it all seems to check out! I really like how this eliminates the issue of platform-specific changes, as well as the name-generating macro. Thanks, @ahelium!

File: misc/dbt-materialize/dbt/adapters/materialize/impl.py (hunk 1 above)
Comment (ahelium) [reply misc/dbt-materialize/dbt/adapters/materialize/impl.py]: Oh, gosh, thanks for pointing this out.

**Pull request merged**